File: pidtrain.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from time import time, sleep
from numpy.random import rand, random_sample

# constants
ENVIONMENT_TEMP = 43


class PID():
    def __init__(self, kp, ki, kd, setpoint, period):
        self.temp = ENVIONMENT_TEMP
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.setpoint = setpoint
        self.period = period
        self.integral = 0
        self.last_error = 0
        self.start_time = time()
        self.curr_cycle = 1

    def get_integral(self):
        return self.integral

    def set_integral(self, value):
        self.integral = value

    def get_last_error(self):
        return self.last_error

    def set_last_error(self, value):
        self.last_error = value

    def get_start_time(self):
        return self.start_time

    def get_setpoint(self):
        return self.setpoint

    def get_period(self):
        return self.period

    def get_pid(self):
        return self.kp, self.ki, self.kd

    def set_temp(self, value):
        self.temp = value

    def get_temp(self):
        return self.temp

    def set_curr_cycle(self, value):
        self.curr_cycle = value

    def get_curr_cycle(self):
        return self.curr_cycle

    def output(self, temp):
        error = self.get_setpoint() - temp
        integral = self.get_integral() + error
        derivative = (error - self.get_last_error()) / self.get_period() / 1e3
        kp, ki, kd = self.get_pid()
        self.set_integral(integral)
        self.set_last_error(error)
        return kp * error + ki * integral + kd * derivative

    def simulation(self, cycle_num):
        output_logger = []
        temp_logger = []
        self.set_curr_cycle(1)
        while cycle_num + 1 > self.get_curr_cycle():
            temp_curr = self.get_temp()
            temp_logger.append(temp_curr)
            output = self.output(temp_curr)
            output_logger.append(output)
            if output > 0:
                self.set_temp(temp_curr + 0.3)
            else:
                self.set_temp(temp_curr - 0.01)
            # print(self.get_curr_cycle(), self.get_temp())
            sleep(self.get_period())
            self.set_curr_cycle(self.get_curr_cycle() + 1)
        return temp_logger, output_logger

    @staticmethod
    def mae_cost(data, setpoint):
        n = len(data)
        return sum([abs(dt - setpoint) for dt in data]) / n

    @staticmethod
    def mse_cost(data, setpoint):
        n = len(data)
        return sum([(dt - setpoint) ** 2 for dt in data]) / n


def pid_hybrid(pid1, pid2):
    dice = rand() * 3
    if 0 <= dice < 1:
        pid1_new = np.array([pid1[0], pid2[1], pid2[2]])
        pid2_new = np.array([pid2[0], pid1[1], pid1[2]])
    elif 1 <= dice < 2:
        pid1_new = np.array([pid2[0], pid1[1], pid2[2]])
        pid2_new = np.array([pid1[0], pid2[1], pid1[2]])
    else:
        pid1_new = np.array([pid2[0], pid2[1], pid1[2]])
        pid2_new = np.array([pid1[0], pid1[1], pid2[2]])
    return pid1_new, pid2_new


def pid_mutation():
    return 100 * random_sample(3)


# Genetic Algorithm Simulation
def genalg_simu(GenMax, PopSize, setpoint, period, cycle_num, is_mae=True):
    pid_data = pd.DataFrame()
    for gen in range(GenMax):
        print('generation # ', gen)
        print('--------------')
        print('Kp, Ki, Kd, Cost')
        # First generation
        if gen == 0:
            K = 100 * random_sample((PopSize, 3))
        Cost = []
        pid_data_gen = pd.DataFrame()
        # run simulation for each
        for i in range(PopSize):
            pid = PID(K[i][0], K[i][1], K[i][2], setpoint, period)
            cost = pid.mae_cost(pid.simulation(cycle_num)[0], setpoint) if is_mae else pid.mse_cost(
                pid.simulation(cycle_num)[0], setpoint)
            Cost.append(cost)
            print(K[i][0], K[i][1], K[i][2], cost)
            pid_data_gen = pid_data_gen.append(
                {'Kp': K[i][0], 'Ki': K[i][1], 'Kd': K[i][2], 'Cost': cost, 'Gen': int(gen)},
                ignore_index=True)
        # Genetic algorithm
        pid_data_gen = pid_data_gen.sort_values(by=['Cost'])
        pid_data = pid_data.append(pid_data_gen)
        # update K matrix
        K[:, 0] = pid_data_gen['Kp'].tolist()
        K[:, 1] = pid_data_gen['Ki'].tolist()
        K[:, 2] = pid_data_gen['Kd'].tolist()
        # Elitism
        K_prime = K
        # hybrid
        for j in np.linspace(1, PopSize - 3, 1, dtype='int'):
            K_prime[j], K_prime[j + 1] = pid_hybrid(K[j], K[j + 1])
        if PopSize % 2:
            K_prime[PopSize - 1] = pid_mutation()  # the last to mutate
        else:
            K_prime[PopSize - 2] = pid_mutation()  # the last two to mutate
            K_prime[PopSize - 1] = pid_mutation()
        K = K_prime
        pid_data.to_csv("simu_setpt_" + str(setpoint) + "_genmax_" + str(GenMax) + "_popsz_" + str(PopSize))
    return pid_data


if __name__ == "__main__":
    setpoint = 65
    period = 0.001
    GenMax = 100
    PopSize = 19
    cycle_num = 1000

    pid_data = genalg_simu(GenMax, PopSize, setpoint, period, cycle_num, True)

    # plot
    fig = plt.figure()
    ax = fig.add_subplot(projection='3d')
    pdata = pid_data['Kp'].tolist()
    idata = pid_data['Ki'].tolist()
    ddata = pid_data['Kd'].tolist()
    gdata = pid_data['Gen'].tolist()
    ax.scatter(pdata, idata, ddata, color='b')
    ax.set_xlim(0, 100)
    ax.set_ylim(0, 100)
    ax.set_zlim(0, 100)
    ax.set_xlabel('Kp')
    ax.set_ylabel('Ki')
    ax.set_zlabel('Kd')
    plt.show()

File: pidtrain.cpp
// pidtrain.cpp
#include<iostream>
#include<ctime> // for time()
#include<vector>
#include<chrono> // for chrono
#include<thread> // for this_thread
#include<numeric> // for accumulate()
#include<cmath> // for abs()
#include<random> // for random_device
#include<algorithm> // for sort()
#include<fstream> // for file I/O
#include<iostream>

#define ENVIRONMENT_TEMP 43
using namespace std;


struct pidcoef // PID coefficients
{
	float Kp;
	float Ki;
	float Kd;
	float cost = 0.0; // cost function value
};

class PID
{
private:
	pidcoef Kpid;
	float temp;
	float setpoint;
	float integral;
	float last_error;
	float start_time;
	float period;
	int curr_cycle;
public:
	PID(float kp, float ki, float kd, float sp, float pd) :
		temp(ENVIRONMENT_TEMP), Kpid({ kp, ki, kd }), setpoint(sp), period(pd), integral(0),
		last_error(0), start_time(std::time(0)), curr_cycle(1)
	{  }; // contructor
	//~PID(); // destructor
	float get_integral() const { return integral; }
	float get_last_error() const { return last_error; }
	float get_start_time() const { return start_time; }
	float get_setpoint() const { return setpoint; }
	float get_temp() const { return temp; }
	float get_curr_cycle() const { return curr_cycle; }
	float get_period() const { return period; }
	pidcoef get_pid() const { return Kpid; }
	void set_curr_cycle(int cc) { curr_cycle = cc; }
	void set_integral(float it) { integral = it; }
	void set_temp(float t) { temp = t; }
	void set_last_error(float le) { last_error = le; }

	float output() // calculate output for one PID cycle
	{
		float error = 0.0;
		float derivative = 0.0;
		error = setpoint - temp;
		integral = get_integral() + error;
		derivative = (error - get_last_error()) / get_period();
		set_integral(integral);
		set_last_error(error);

		return Kpid.Kp * error + Kpid.Ki * integral + Kpid.Kd * derivative;
	}
	float simulation(vector<float>* temp_logger, int cycle_num) // simulate for cycle_num PID cycles
	{
		set_curr_cycle(1);
		int n = 1;
		while (cycle_num + 1 > curr_cycle)
		{
			temp_logger->push_back(temp);
			if (output() > 0)
				set_temp(temp + 0.3);
			else
				set_temp(temp - 0.01);

			set_curr_cycle(++n);
		}
		vector<float> tempVtr;
		for (int i = 0; i < temp_logger->size(); i++)
			tempVtr.push_back(abs((*temp_logger)[i] - setpoint));
		float sum = accumulate(tempVtr.begin(), tempVtr.end(), 0);
		tempVtr.clear();
		return sum / temp_logger->size();
	};

};

void swap(float& a, float& b) // swap the values of a and b
{
	float temp;
	temp = a;
	a = b;
	b = temp;
}

void pid_hybrid(pidcoef& Kpid1, pidcoef& Kpid2) // hybrid two PID parameters and initialize
{
	int dice = (rand() % 3);
	switch (dice)
	{
	case 0:
	{
		swap(Kpid1.Ki, Kpid2.Ki);
		swap(Kpid1.Kd, Kpid2.Kd);
		Kpid1.cost = 0;
		Kpid2.cost = 0;
		break;
	}
	case 1:
	{
		swap(Kpid1.Kp, Kpid2.Kp);
		swap(Kpid1.Kd, Kpid2.Kd);
		Kpid1.cost = 0;
		Kpid2.cost = 0;
		break;
	}
	case 2:
	{
		swap(Kpid1.Ki, Kpid2.Ki);
		swap(Kpid1.Kp, Kpid2.Kp);
		Kpid1.cost = 0;
		Kpid2.cost = 0;
		break;
	}
	}

}


pidcoef pid_mutation() // generate random PID coefficients to realize mutation
{
	random_device rd; // obtain a random number from hardware
	mt19937 gen(rd()); // seed the generator
	uniform_int_distribution<> distr(0, 100); // define the range
	pidcoef Kpid;
	Kpid.Kp = distr(gen);
	Kpid.Ki = distr(gen);
	Kpid.Kd = distr(gen);
	Kpid.cost = 0;
	return Kpid;
}

void fill_random(vector<pidcoef>* vtr, int size) // fill a vector data with random float numbers
{
	random_device rd;
	mt19937 gen(rd());
	uniform_real_distribution<float> distr(0, 100);

	for (int i = 0; i < size; i++)
	{
		pidcoef new_kpid;
		new_kpid.Kp = distr(gen);
		new_kpid.Ki = distr(gen);
		new_kpid.Kd = distr(gen);
		vtr->push_back(new_kpid);
	};
};

bool cost_sorter(pidcoef const& Kpid1, pidcoef const& Kpid2) // sort by cost function value
{
	return Kpid1.cost < Kpid2.cost;
}


void genalg_simu(int GenMax, int PopSize, float setpoint, float period, int cycle_num) // For simulation
{
	vector<pidcoef> Kpid_vtr;
	fill_random(&Kpid_vtr, PopSize);
	for (int gen = 0; gen < GenMax; gen++)
	{
		cout << "\ngeneration # " << gen << endl;
		cout << "-------------\n" << endl;
		cout << "Kp, Ki, Kd, Cost \n" << endl;


		for (int i = 0; i < PopSize; i++)
		{
			PID pid(Kpid_vtr[i].Kp, Kpid_vtr[i].Ki, Kpid_vtr[i].Kd, setpoint, period);
			vector<float> temp_logger;
			Kpid_vtr[i].cost = pid.simulation(&temp_logger, cycle_num);


			cout << Kpid_vtr[i].Kp << ' ' << Kpid_vtr[i].Ki << ' ' << Kpid_vtr[i].Kd
				<< ' ' << Kpid_vtr[i].cost << '\n';
		};

		sort(Kpid_vtr.begin(), Kpid_vtr.end(), &cost_sorter);
		for (int j = 1; j < PopSize - 2; j += 2)
			pid_hybrid(Kpid_vtr[j], Kpid_vtr[j + 1]);
		if (PopSize % 2)
			Kpid_vtr[PopSize - 1] = pid_mutation();
		else
		{
			Kpid_vtr[PopSize - 2] = pid_mutation();
			Kpid_vtr[PopSize - 1] = pid_mutation();
		}
	};
};



int main()
{
	auto t1 = std::chrono::high_resolution_clock::now();
	float setpoint = 65;
	float period = 1;
	int GenMax = 100;
	int PopSize = 20;
	int cycle_num = 1000;
	genalg_simu(GenMax, PopSize, setpoint, period, cycle_num);
	auto t2 = std::chrono::high_resolution_clock::now();
	auto ms_int = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1);
	std::cout << ms_int.count() << "ms\n"; // display runtime
	return 0;
}